### ISSUE TITLE
Multi thread xunit tests randomly fails

### ISSUE THREAD

**Issue by hikalkan** [zzzprojects/EntityFramework-Effort#48]
Hi,

Some of our unit tests randomly fails because of the following exception:

```
System.Data.Entity.Infrastructure.CommitFailedException : An error was reported while committing a database transaction but it could not be determined whether the transaction succeeded or failed on the database server. See the inner exception and http://go.microsoft.com/fwlink/?LinkId=313468 for more information.
---- System.Transactions.TransactionAbortedException : The transaction has aborted.
-------- System.TimeoutException : Transaction Timeout
```

When we run it single threaded, it works fine. Also, most of times all unit tests works normally, but sometimes it throws that exception.

Do you have an idea?
Thanks.

**Comment by tamasflamich** [comment]
Hi hikalkan - which kind of factory method do you use? transient or persistent?

**Comment by floyd-may** [comment]
It's using a transient connection, and sharing the resulting `DbConnection` as a singleton in the code under test (via Castle Windsor).

**Comment by tamasflamich** [comment]
That's seems to be quite unusual. Is there any chance to determine the stacktrace where the application is actually stuck?

**Comment by floyd-may** [comment]
In many cases, it's at a DbContext's `SaveChanges()` method. For instance:
```
System.Data.Entity.Infrastructure.CommitFailedException : An error was reported while committing a database transaction but it could not be determined whether the transaction succeeded or failed on the database server. See the inner exception and http://go.microsoft.com/fwlink/?LinkId=313468 for more information.
---- System.Transactions.TransactionAbortedException : The transaction has aborted.
-------- System.TimeoutException : Transaction Timeout
Stack Trace:
(TTarget target, Action`2 operation, TInterceptionContext interceptionContext, Action`3 executing, Action`3 executed)
   at System.Data.Entity.Infrastructure.Interception.DbTransactionDispatcher.Commit(DbTransaction transaction, DbInterceptionContext interceptionContext)
   at System.Data.Entity.Core.EntityClient.EntityTransaction.Commit()
(Func`1 func, IDbExecutionStrategy executionStrategy, Boolean startLocalTransaction, Boolean releaseConnectionOnSuccess)
   at System.Data.Entity.Core.Objects.ObjectContext.SaveChangesToStore(SaveOptions options, IDbExecutionStrategy executionStrategy, Boolean startLocalTransaction)
   at System.Data.Entity.Core.Objects.ObjectContext.<>c__DisplayClass2a.<SaveChangesInternal>b__27()
(Func`1 operation)
   at System.Data.Entity.Core.Objects.ObjectContext.SaveChangesInternal(SaveOptions options, Boolean executeInExistingTransaction)
   at System.Data.Entity.Core.Objects.ObjectContext.SaveChanges(SaveOptions options)
   at System.Data.Entity.Internal.InternalContext.SaveChanges()
   at System.Data.Entity.Internal.LazyInternalContext.SaveChanges()
   at Abp.EntityFramework.AbpDbContext.SaveChanges()
   <our application-specific code>
----- Inner Stack Trace -----
   at System.Transactions.TransactionStateAborted.BeginCommit(InternalTransaction tx, Boolean asyncCommit, AsyncCallback asyncCallback, Object asyncState)
   at System.Transactions.CommittableTransaction.Commit()
   at Effort.Provider.EffortTransaction.Commit()
   at System.Data.Entity.Infrastructure.Interception.DbTransactionDispatcher.<Commit>b__c(DbTransaction t, DbTransactionInterceptionContext c)
(TTarget target, Action`2 operation, TInterceptionContext interceptionContext, Action`3 executing, Action`3 executed)
----- Inner Stack Trace -----
```

The ABP framework uses `TransactionScope`s rather than `DbConnection.BeginTransaction(...)`, if that's any help.

**Comment by krugerm** [comment]
+1

**Comment by floyd-may** [comment]
@tamasflamich Any recommendations?

**Comment by tamasflamich** [comment]
Unfortunately, this stack trace doesn't really help a lot. I am not familiar with the ABP framework, so I wonder what it exactly does to .NET's translation library? The transaction handling of NMemory heavily relies on it, so it is possible that they are not that compatible with each other.

**Comment by floyd-may** [comment]
@tamasflamich  Sorry, I didn't quite understand that. Could you maybe explain in a bit more detail?

**Comment by Mr-hai** [comment]
Hi . @tamasflamich 

This stack trace does help a lot 
```
----- Inner Stack Trace -----
             at NMemory.Transactions.Transaction.EnsureTransaction(Transaction& transaction, IDatabase database)
             at NMemory.Tables.Table`2.Update(IQueryable`1 query, IUpdater`1 updater, Transaction transaction)
             at NMemory.Tables.Table`2.NMemory.Tables.IBulkTable<TEntity>.Update(TableQuery`1 query, Expression`1 updater, Transaction transaction)
             at NMemory.Linq.QueryableEx.Update[T](IQueryable`1 queryable, Expression`1 updater, Transaction transaction)
             at Effort.Internal.Common.DatabaseReflectionHelper.WrapperMethods.UpdateEntities[TEntity](IQueryable`1 query, Expression`1 updater, Transaction transaction)
```

**Comment by prateek14** [comment]
+1

**Comment by raphaeldiasrj** [comment]
+1

**Comment by raphaeldiasrj** [comment]
@tamasflamich Maybe this stacktrace helps you

```
System.Data.Entity.Core.UpdateException
An error occurred while updating the entries. See the inner exception for details.
   em System.Data.Entity.Core.Mapping.Update.Internal.UpdateTranslator.Update()
   em System.Data.Entity.Core.Objects.ObjectContext.ExecuteInTransaction[T](Func`1 func, IDbExecutionStrategy executionStrategy, Boolean startLocalTransaction, Boolean releaseConnectionOnSuccess)
   em System.Data.Entity.Core.Objects.ObjectContext.SaveChangesToStore(SaveOptions options, IDbExecutionStrategy executionStrategy, Boolean startLocalTransaction)
   em System.Data.Entity.Core.Objects.ObjectContext.SaveChangesInternal(SaveOptions options, Boolean executeInExistingTransaction)
   em System.Data.Entity.Internal.InternalContext.SaveChanges()

System.Reflection.TargetInvocationException
Uma exceção foi acionada pelo destino de uma chamada.
   em System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor)
   em System.Reflection.RuntimeMethodInfo.UnsafeInvokeInternal(Object obj, Object[] parameters, Object[] arguments)
   em System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   em Effort.Internal.Common.DatabaseReflectionHelper.InsertEntity(ITable table, Object entity, Transaction transaction)
   em Effort.Internal.CommandActions.InsertCommandAction.CreateAndInsertEntity(ITable table, IList`1 memberBindings, Transaction transaction)
   em Effort.Internal.CommandActions.InsertCommandAction.ExecuteNonQuery(ActionContext context)
   em System.Data.Entity.Infrastructure.Interception.InternalDispatcher`1.Dispatch[TTarget,TInterceptionContext,TResult](TTarget target, Func`3 operation, TInterceptionContext interceptionContext, Action`3 executing, Action`3 executed)
   em System.Data.Entity.Infrastructure.Interception.DbCommandDispatcher.NonQuery(DbCommand command, DbCommandInterceptionContext interceptionContext)
   em System.Data.Entity.Core.Mapping.Update.Internal.DynamicUpdateCommand.Execute(Dictionary`2 identifierValues, List`1 generatedValues)
   em System.Data.Entity.Core.Mapping.Update.Internal.UpdateTranslator.Update()

System.Transactions.TransactionAbortedException
A transação foi anulada.
   em NMemory.Transactions.Transaction.EnsureTransaction(Transaction& transaction, IDatabase database)
   em NMemory.Tables.Table`2.Insert(TEntity entity, Transaction transaction)
```

Any sugestion?

**Comment by joel-cass** [comment]
I've run into the same issue - here's my hunch:

EffortProviderConfiguration.RegisterProvider(); << STATIC METHOD
var connection = DbConnectionFactory.CreateTransient(); << STATIC METHOD

It looks like the provider is creating a database that matches the instanceId and marking it as transient; however only one connection seems to be able to be transient at a time. It would be great to either:
a) get rid of the static methods and instead use a Provider instance that encloses all its data for easy use/transfer/disposal.
b) allow a transient connection to be specific to a thread instead of instance.

**Comment by brase** [comment]
I am stumbled over the same Type of error with the same exception:

```
SetUp : System.Exception : Beim Herstellen der Verbindung zur Datenbank ist ein Fehler aufgetreten.
  ----> System.Data.UpdateException : Fehler beim Aktualisieren der Einträge. Weitere Informationen finden Sie in der internen Ausnahme.
  ----> System.Reflection.TargetInvocationException : Ein Aufrufziel hat einen Ausnahmefehler verursacht.
  ----> System.Transactions.TransactionAbortedException : Die Transaktion wurde abgebrochen.
   bei MediFoxRZ.Data.Mapping.ContextFactory.CreateContext() in D:\BFS\BfsFinance\DataAccess\MediFoxRZ.Data.Mapping\ContextFactory.cs:Zeile 28.
   bei MediFoxRZ.Services.Common.ServiceBase.Get[T](Func`2 func) in D:\BFS\BfsFinance\Services\MediFoxRZ.Services.Common\ServiceBase.cs:Zeile 15.
   bei MediFoxRZ.Services.Abrechnung.AbrechnungService.GetNachberechnungsgruende() in D:\BFS\BfsFinance\Services\MediFoxRZ.Services.Abrechnung\AbrechnungService.cs:Zeile 2571.
   bei MediFoxRZ.Services.Abrechnung.Tests.NachberechnungTests.Wenn_man_nach_existierenden_Nachberechnungsgruenden_fragt.Mit_Demo_Daten() in D:\BFS\BfsFinance\Services\MediFoxRZ.Services.Abrechnung.Tests\NachberechnungTests.cs:Zeile 30.
--UpdateException
   bei System.Data.Mapping.Update.Internal.UpdateTranslator.Update(IEntityStateManager stateManager, IEntityAdapter adapter)
   bei System.Data.Objects.ObjectContext.SaveChanges(SaveOptions options)
   bei MediFoxRZ.Services.Tests.With_InMemory_Db.<With_initialized_Entities>b__3_0(String constring) in D:\BFS\BfsFinance\Services\MediFoxRZ.Services.Tests\With_InMemory_Db.cs:Zeile 40.
   bei MediFoxRZ.Data.Mapping.ContextFactory.CreateContext() in D:\BFS\BfsFinance\DataAccess\MediFoxRZ.Data.Mapping\ContextFactory.cs:Zeile 24.
--TargetInvocationException
   bei System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor)
   bei System.Reflection.RuntimeMethodInfo.UnsafeInvokeInternal(Object obj, Object[] parameters, Object[] arguments)
   bei System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   bei System.Reflection.MethodBase.Invoke(Object obj, Object[] parameters)
   bei Effort.Internal.Common.DatabaseReflectionHelper.InsertEntity(ITable table, Object entity, Transaction transaction) in C:\Users\brase\code\rs\spielwiese\effort\Main\Source\Effort\Internal\Common\DatabaseReflectionHelper.cs:Zeile 125.
   bei Effort.Internal.CommandActions.InsertCommandAction.CreateAndInsertEntity(ITable table, IList`1 memberBindings, Transaction transaction) in C:\Users\brase\code\rs\spielwiese\effort\Main\Source\Effort\Internal\CommandActions\InsertCommandAction.cs:Zeile 159.
   bei Effort.Internal.CommandActions.InsertCommandAction.ExecuteNonQuery(ActionContext context) in C:\Users\brase\code\rs\spielwiese\effort\Main\Source\Effort\Internal\CommandActions\InsertCommandAction.cs:Zeile 137.
   bei Effort.Provider.EffortEntityCommand.ExecuteNonQuery() in C:\Users\brase\code\rs\spielwiese\effort\Main\Source\Effort\Provider\EffortEntityCommand.cs:Zeile 92.
   bei System.Data.Mapping.Update.Internal.DynamicUpdateCommand.Execute(UpdateTranslator translator, EntityConnection connection, Dictionary`2 identifierValues, List`1 generatedValues)
   bei System.Data.Mapping.Update.Internal.UpdateTranslator.Update(IEntityStateManager stateManager, IEntityAdapter adapter)
--TransactionAbortedException
   bei NMemory.Transactions.Transaction.EnsureTransaction(Transaction& transaction, IDatabase database)
   bei NMemory.Tables.Table`2.Insert(TEntity entity, Transaction transaction)
   bei Effort.Internal.Common.DatabaseReflectionHelper.WrapperMethods.InsertEntity[TEntity](ITable`1 table, TEntity entity, Transaction transaction) in C:\Users\brase\code\rs\spielwiese\effort\Main\Source\Effort\Internal\Common\DatabaseReflectionHelper.cs:Zeile 285.
```

**Comment by latebrose** [comment]
I don't why be it worked.

change
```
for(var item in items){
    context.Items.Add(item);
}
context.saveChanges();
```
to
```
for(var item in items){
    context.Items.Add(item);
    context.saveChanges();
}
```

**Comment by JonathanMagnan** [comment]
Hello @latebrose ,

Do you share the same context between unit test?

Do you think you could provide a "working" project example/test with this issue?

It will make easier/faster for my developer to getting started for investigating it.

_You can send it to info@zzzprojects.com if you need to keep the source private_

_We now always ask for a project since we found out that most issues are missing some essential information, have some syntax mistake or are resolved by the requestor when creating it_

_(Even if the issue seem very easy to reproduce, by getting a test project, it allow us to give a faster support and better experience for the support of all our free libraries)_

Best Regards,

Jonathan

---

**Performance Libraries**
`context.BulkInsert(list, options => options.BatchSize = 1000);  `
[Entity Framework Extensions](http://entityframework-extensions.net/) • [Bulk Operations](http://bulk-operations.net/) •  [Dapper Plus](http://dapper-plus.net/) •  [LinqToSql Plus](http://linqtosql-plus.net/)

**Runtime Evaluation**
`Eval.Execute("x + y", new {x = 1, y = 2}); // return 3`
[C# Eval Function](http://eval-expression.net/) • [SQL Eval Function](http://eval-sql.net/)

**Comment by JonathanMagnan** [comment]
Hello,

This issue will be closed since none provided information.

We understand this issue has been created long ago, but since that's a free library, we want to make sure to work on an issue that is still active since we cannot make them all.

Feel free to reopen if you can provide further information.

Best Regards,

Jonathan

**Comment by mgolubovic82** [comment]
I'm reporting my findings here in case anyone runs into this and has the same problem I did.  

In my case it had to do with the save transaction timing out in some cases, but not others.  The exception was not clear that this was happening, it was just a transaction aborted exception as the inner most exception.   I was able to get around it by setting the Effort connection to a large number which is available in 2.0.5 I believe.   

 var connection = Effort.DbConnectionFactory.CreateTransient();
            connection.SetConnectionTimeout(500000);
